# Foundations of Data Science Project - Diabetes Analysis

---------------
## Context
---------------

Diabetes is one of the most frequent diseases worldwide and the number of diabetic patients are growing over the years. The main cause of diabetes remains unknown, yet scientists believe that both genetic factors and environmental lifestyle play a major role in diabetes.

A few years ago research was done on a tribe in America which is called the Pima tribe (also known as the Pima Indians). In this tribe, it was found that the ladies are prone to diabetes very early. Several constraints were placed on the selection of these instances from a larger database. In particular, all patients were females at least 21 years old of Pima Indian heritage. 

-----------------
## Objective
-----------------

Here, we are analyzing different aspects of Diabetes in the Pima Indians tribe by doing Exploratory Data Analysis.

-------------------------
## Data Dictionary
-------------------------

The dataset has the following information:

* Pregnancies: Number of times pregnant
* Glucose: Plasma glucose concentration over 2 hours in an oral glucose tolerance test
* BloodPressure: Diastolic blood pressure (mm Hg)
* SkinThickness: Triceps skin fold thickness (mm)
* Insulin: 2-Hour serum insulin (mu U/ml)
* BMI: Body mass index (weight in kg/(height in m)^2)
* DiabetesPedigreeFunction: A function which scores likelihood of diabetes based on family history.
* Age: Age in years
* Outcome : Class variable (0: person is not diabetic or 1: person is diabetic)

## Q 1: Import the necessary libraries and briefly explain the use of each library (3 Marks)

In [ ]:
#remove _____ & write the appropriate library name

import numpy as np
import pandas as pd

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

#### Write your Answer here: 

Ans 1:

- numpy: 
  - NumPy is a python package that **stands for ‘Numerical Python’**. It is a python package for consolidating the **handling of numbers** on numerical analysis or numerical methoods.

  - NumPy is for when we are dealing with **numbers, instead of data**.

  - Numpy is the core library for scientific computing, which contains a powerful n-dimensional array object, provides tools for integrating C, C++, etc.

- pandas:
  - Pandas contains data structures and data manipulation tools designed for **data cleaning and analysis**.

  - While pandas adopt many coding idioms from NumPy, the biggest difference is that **pandas is designed for working with tabular or heterogeneous data**. NumPy, by contrast, is best suited for working with homogeneous numerical array data.

  - The name Pandas is derived from the term **“panel data”**, an econometrics term for multidimensional structured data sets.

- seaborn:
  - Seaborn is a Python **visualization library based on matplotlib**. The library provides a high-level interface **for plotting statistical graphics**. As the library uses matplotlib in the backend, **we can use the functions in matplotlib along with functions in seaborn**.

  - Various functions in the seaborn library allow us to **plot complex and advance statistical plots** like linear/higher-order regression, univariate/multivariate distribution, violin, swarm, strip plots, correlations and so on.

- matplotlib:
  - Matplotlib is a Python **2D plotting library**. Many libraries are built on top of it and use its functions in the backend.

- pyplot:
  - pyplot is a **subpackage of matplotlib** that provides a MATLAB-like way of **plotting**.

  - matplotlib.pyplot is a mostly used package because it is very simple to use and it generates plots in less time.

## Q 2: Read the given dataset (1 Mark)

In [4]:
# remove _____ & write the appropriate function name

pima = pd.read_csv("./data/diabetes.csv")

## Q3. Show the last 10 records of the dataset. How many columns are there? (1 Mark)

In [7]:
#remove ______ and write the appropriate number in the function

print('Number of columns :', pima.shape[1])

pima.tail(10)

Number of columns : 9


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
758            1      106             76              0        0  37.5   
759            6      190             92              0        0  35.5   
760            2       88             58             26       16  28.4   
761            9      170             74             31        0  44.0   
762            9       89             62              0        0  22.5   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
758                     0.197   26        0  
759                     0.278   66        1  
760                     0.766   22        0  
761                     0.403   43        1  
762                     0.142   33        0  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

#### Write your Answer here: 


## Q4. Show the first 10 records of the dataset (1 Mark)

In [8]:
#remove _____ & write the appropriate function name and the number of rows to get in the output

pima.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
5                     0.201   30        0  
6                     0.248   26        1  
7                     0.134   29        0  
8                     0.158   53        1  
9                     0.232   54        1

## Q5. What do you understand by the dimension of the dataset? Find the dimension of the `pima` dataframe. (1 Mark)

In [20]:
#remove _____ & write the appropriate function name

print('The shape of the dataset is: ',pima.shape)

The shape of the dataset is:  (768, 9)


#### Write your Answer here: 


Ans 5:
- Each *unique* feature of the dataset is a dimention. Since our dataset has 9 features, it is a **9D dataset**.

## Q6. What do you understand by the size of the dataset? Find the size of the `pima` dataframe. (1 Mark)

In [21]:
#remove _____ & write the appropriate function name

print('The size of the dataset is: ',pima.size)

The size of the dataset is:  6912


#### Write your Answer here: 


Ans 6:
- The size of the dataset is the result of the multiplication of the number of rows by the number of columns.

## Q7. What are the data types of all the variables in the data set? (2 Marks)
**Hint: Use info() function to get all the information about the dataset.**

In [22]:
#remove _____ & write the appropriate function name

pima.dtypes

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

#### Write your Answer here: 


Ans 7:
- Pregnancies: integer
- Glucose: integer
- BloodPressure: integer
- SkinThickness: integer
- Insulin: integer
- BMI: *floating point number*
- DiabetesPedigreeFunction: *floating point number*
- Age: integer
- Outcome: integer

## Q8. What do you mean by missing values? Are there any missing values in the `pima` dataframe? (2 Marks)

In [29]:
#remove _____ & write the appropriate function name

def missing_check(df):
    total = df.isnull().sum().sort_values(ascending=False)   # total number of null values
    percent = (df.isnull().sum()/df.isnull().count()).sort_values(ascending=False)  # percentage of values that are null
    missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])  # putting the above two together
    return missing_data # return the dataframe
missing_check(pima)

Total  Percent
Pregnancies                   0      0.0
Glucose                       0      0.0
BloodPressure                 0      0.0
SkinThickness                 0      0.0
Insulin                       0      0.0
BMI                           0      0.0
DiabetesPedigreeFunction      0      0.0
Age                           0      0.0
Outcome                       0      0.0

#### Write your Answer here: 


Ans 8:
- We may come across datasets with incomplete data, this means that there may be some values missing from the dataset.
- As we can see in the table above, there are no missing values in this dataframe.

## Q9. What does summary statistics of data represents? Find the summary statistics for all variables except 'Outcome' in the `pima` data? Take one column/variable from the output table and explain all the statistical measures. (3 Marks)

In [27]:
#remove _____ & write the appropriate function name

pima.iloc[:,0:8].describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000                  0.243750   24.000000  
50%     32.000000                  0.372500   29.000000  
75%     36.600000                  0.626250   41.000000  
max     67.100000                  2.420000   81.000000

#### Write your Answer here: 


Ans 9:
- Summary statistics *sumarize* and provide information about our sample data, they help us measure the location of our data, trends in our data, spread, skewness, etc. THey help us paint an initial picture of our data so we can decide how to analyze it moving forward.
- For example, if we look at the 'Pregnancies' column in the table above we can learn a couple of things:
  - Count:
    - We are analyzing the data of 768 individuals, since the count returns the amount of cells in the pregnancy column, regardless of their value.
  - Mean:
    - The average number of pregnancies in our sample is ~3.8 (we can round up to 4)
  - STD(standard deviation):
    - The STD in our dataset is ~3.37, which means that 68% (roughly two thirds) of the individuals in our dataset had between 0 and 7 pregnancies.
  - Min:
    - The lowest amount of pregnancies for a given individual in our dataset is 0(no pregnancies). Note that this value can be an outlier and it is not representative of the whole dataset.
  - 25%:
    - 75% of the individuals in our dataset had at least 1 pregnancy
  - 50%:
    - 50% of the individuals in our dataset had at least 3 pregnancies
  - 75%:
    - Only 75% of the individuals in our dataset had more than 6 pregnancies
  - Max:
    - The individual with the highest amount of pregnancies in our sample had 17 pregnancies. Note that this value can be an outlier and it is not representative of the whole dataset.

## Q 10. Plot the distribution plot for the variable 'BloodPressure'. Write detailed observations from the plot. (2 Marks)

In [ ]:
#remove _____ & write the appropriate library name

______.displot(pima['BloodPressure'], kind='kde')
plt.show()

#### Write your Answer here: 


## Q 11. What is the 'BMI' for the person having the highest 'Glucose'? (1 Mark)

In [ ]:
#remove _____ & write the appropriate function name

pima[pima['Glucose']==pima['Glucose']._____()]['BMI']

#### Write your Answer here: 


## Q12.
### 12.1 What is the mean of the variable 'BMI'? 
### 12.2 What is the median of the variable 'BMI'? 
### 12.3 What is the mode of the variable 'BMI'?
### 12.4 Are the three measures of central tendency equal?

### (3 Marks)

In [ ]:
#remove _____ & write the appropriate function name

m1 = pima['BMI'].____()  #Mean
print(m1)
m2 = pima['BMI'].____()  #Median
print(m2)
m3 = pima['BMI'].____()[0]  #Mode
print(m3)

#### Write your Answer here: 


## Q13. How many women's 'Glucose' level is above the mean level of 'Glucose'? (1 Mark)

In [ ]:
#remove _____ & write the appropriate function name

pima[pima['Glucose']>pima['Glucose']._____()].shape[0]

#### Write your Answer here: 


## Q14. How many women have their 'BloodPressure' equal to the median of 'BloodPressure' and their 'BMI' less than the median of 'BMI'? (2 Marks)

In [ ]:
#remove _____ & write the appropriate column name

pima[(pima['BloodPressure']==pima['BloodPressure'].______()) & (pima['BMI']<pima['BMI']._______())]

#### Write your Answer here: 


## Q15. Create the pairplot for variables 'Glucose', 'SkinThickness' and 'DiabetesPedigreeFunction'. Write you observations from the plot. (4 Marks)

In [ ]:
#remove _____ & write the appropriate function name

sns.______(data=pima,vars=['Glucose', 'SkinThickness', 'DiabetesPedigreeFunction'], hue='Outcome')
plt.show()

#### Write your Answer here: 


## Q16. Plot the scatterplot between 'Glucose' and 'Insulin'. Write your observations from the plot. (2 Marks)

In [ ]:
#remove _____ & write the appropriate function name

sns._______(x='Glucose',y='Insulin',data=pima)
plt.show(

#### Write your Answer here: 


## Q 17. Plot the boxplot for the 'Age' variable. Are there outliers? (2 Marks)

In [ ]:
#remove _____ & write the appropriate function and column name 

plt.______(pima['____'])

plt.title('Boxplot of Age')
plt.ylabel('Age')
plt.show()

#### Write your Answer here: 


## Q18. Plot histograms for variable Age to understand the number of women in different Age groups given that they have diabetes or not. Explain both histograms and compare them. (3 Marks)

In [ ]:
#remove _____ & write the appropriate function and column name

plt._____(pima[pima['Outcome']==1]['_____'], bins = 5)
plt.title('Distribution of Age for Women who has Diabetes')
plt.xlabel('Age')
plt.ylabel('Frequency')
plt.show()

In [ ]:
#remove _____ & write the appropriate function and column name

plt.______(pima[pima['Outcome']==0]['_____'], bins = 5)
plt.title('Distribution of Age for Women who do not have Diabetes')
plt.xlabel('Age')
plt.ylabel('Frequency')
plt.show()

#### Write your Answer here: 


## Q 19. What is Inter Quartile Range of all the variables? Why is it used? Which plot visualizes the same? (2 Marks)

In [ ]:
#remove _____ & write the appropriate variable name

Q1 = pima.quantile(0.25)
Q3 = pima.quantile(0.75)
IQR = __ - __
print(IQR)

#### Write your Answer here: 


## Q 20. Find and visualize the the correlation matrix. Write your observations from the plot. (3 Marks)

In [ ]:
#remove _____ & write the appropriate function name and run the code

corr_matrix = pima.iloc[:,0:8].______

corr_matrix

In [ ]:
#remove _____ & write the appropriate function name

plt.figure(figsize=(8,8))
sns._______(corr_matrix, annot = True)

# display the plot
plt.show()

#### Write your Answer here: 
